lib/SecretsCrypto/sf-crypto-key.c

```diff
@@ -0,0 +1,833 @@
+#include "sf-crypto-key.h"
+#include "sf-crypto-key-private.h"
+#include "sf-crypto-manager.h"
+#include "sf-common-private.h"
+
+enum SfCryptoKeyProperties {
+    PROP_NAME = 1,
+    PROP_COLLECTION_NAME,
+    PROP_PLUGIN_NAME,
+    PROP_ORIGIN,
+    PROP_ALGORITHM,
+    PROP_OPERATIONS,
+    PROP_CONSTRAINTS,
+    PROP_KEY_SIZE,
+    PROP_PUBLIC_KEY,
+    PROP_PRIVATE_KEY,
+    PROP_SECRET_KEY,
+    PROP_CUSTOM_PARAMS,
+    PROP_FILTER_DATA,
+    PROP_COUNT = PROP_FILTER_DATA
+};
+
+typedef struct SfCryptoKeyPrivate_ SfCryptoKeyPrivate;
+
+struct SfCryptoKeyPrivate_
+{
+    gchar *name;
+    gchar *collection_name;
+    gchar *plugin_name;
+    SfCryptoKeyOrigin origin;
+    SfCryptoAlgorithm algorithm;
+    SfCryptoOperation operations;
+    SfCryptoKeyConstraint constraints;
+    gint key_size;
+    GBytes *public_key;
+    GBytes *private_key;
+    GBytes *secret_key;
+    GPtrArray *custom_params;
+    GHashTable *filter_data;
+};
+
+G_DEFINE_TYPE_WITH_CODE(SfCryptoKey, sf_crypto_key, G_TYPE_INITIALLY_UNOWNED,
+        G_ADD_PRIVATE(SfCryptoKey))
+
+static void _sf_crypto_key_finalize(GObject *object)
+{
+    SfCryptoKey *key = SF_CRYPTO_KEY(object);
+    SfCryptoKeyPrivate *priv = sf_crypto_key_get_instance_private(key);
+
+    if (priv->name)
+        g_free(priv->name);
```

Comment:
I chose to use if() mainly for consistency with the _unref()s, where the null checks are necessary. Also to me this gives a hint that it is perfectly valid for the field to be NULL.

Reply:
Speaking of NULLs, I personally like to make _ref() and _unref() calls NULL tolerant. Which I guess is influenced by Objective C experience where sending retain or release (or any other message) to a nil object was a nop. But that's an arbitrary choice, what's important is consistency and the rest can be learnt.

Reply:
I'd also prefer that, but this seems to be the choice of glib, and when in rome... Also even the g_free seems to flag the if() as G_UNLIKELY, which would indicate that it should be more of a rare occurrence rather than a valid use case.